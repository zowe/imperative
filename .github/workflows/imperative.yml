# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: Imperative CI

on: [push, pull_request]

jobs:
  test:
    if: (github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository) && !contains(github.event.head_commit.message, '[ci skip]')
    runs-on: ${{ matrix.os }}
    outputs:
      npm-resolutions: ${{ steps.npm-update.outputs.result }}

    strategy:
      fail-fast: false
      matrix:
        node-version: [16.x, 18.x]
        os: [windows-2019, ubuntu-latest, macos-latest]

    env:
      OS: ${{ matrix.os }}
      NODE: ${{ matrix.node-version }}

    timeout-minutes: 60

    concurrency:
      group: ${{ matrix.os }}-node-${{ matrix.node-version }}-ci-${{ github.ref }}
      cancel-in-progress: true

    steps:
    - name: Checkout
      uses: actions/checkout@v3

    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v3
      with:
        node-version: ${{ matrix.node-version }}
        check-latest: true

    - name: Disable Lint Annotations
      run: |
        echo "::remove-matcher owner=eslint-compact::"
        echo "::remove-matcher owner=eslint-stylish::"

    - name: Install Dependencies
      run: npm ci

    - name: Update Dependencies
      id: npm-update
      uses: zowe-actions/octorelease/script@v1
      with:
        script: npmUpdate

    - name: Build Source
      id: build
      run: npm run build

    - name: Unit Tests
      if: ${{ always() && steps.build.outcome == 'success' }}
      run: npm run test:unit >> file.txt

    - name: Unlock Keyring
      if: ${{ always() && steps.build.outcome == 'success' && matrix.os == 'ubuntu-latest' }}
      uses: t1m0thyj/unlock-keyring@master

    - name: Integration Tests
      if: ${{ always() && steps.build.outcome == 'success' }}
      run: npm run test:integration >> file.txt

    - name: Archive Results
      if: ${{ always() && steps.build.outcome == 'success' }}
      uses: actions/upload-artifact@v3
      with:
        name: ${{ matrix.os }}-${{ matrix.node-version }}-results
        path: __tests__/__results__/

    - name: Upload Results to Codecov
      if: ${{ always() && steps.build.outcome == 'success' }}
      uses: codecov/codecov-action@v2
      with:
        env_vars: OS,NODE

  release:
    if: github.event_name == 'push' && github.ref_protected
    needs: test
    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v3
      with:
        fetch-depth: 0
        persist-credentials: false
        ref: ${{ github.ref }}

    - name: Use Node.js LTS
      uses: actions/setup-node@v3
      with:
        node-version: 'lts/*'

    - name: Install Dependencies
      run: npm ci

    - name: Update Dependencies
      uses: zowe-actions/octorelease/script@v1
      env:
        GIT_COMMITTER_NAME: ${{ secrets.ZOWE_ROBOT_USER }}
        GIT_COMMITTER_EMAIL: ${{ secrets.ZOWE_ROBOT_EMAIL }}
        NPM_RESOLUTIONS: ${{ needs.test.outputs.npm-resolutions }}
      with:
        script: npmUpdate

    - name: Build Source
      run: npm run build

    - uses: zowe-actions/octorelease@v1
      env:
        GIT_COMMITTER_NAME: ${{ secrets.ZOWE_ROBOT_USER }}
        GIT_COMMITTER_EMAIL: ${{ secrets.ZOWE_ROBOT_EMAIL }}
        GIT_CREDENTIALS: x-access-token:${{ secrets.ZOWE_ROBOT_TOKEN }}
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        NPM_USERNAME: ${{ secrets.ARTIFACTORY_USERNAME }}
        NPM_PASSWORD: ${{ secrets.ARTIFACTORY_PASSWORD }}
        NPM_EMAIL: ${{ secrets.ZOWE_ROBOT_EMAIL }}
