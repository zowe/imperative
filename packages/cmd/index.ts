/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/doc/profiles/definition/ICommandProfileSchema";
export * from "./src/doc/profiles/definition/ICommandProfileProperty";
export * from "./src/doc/profiles/definition/ICommandProfileTypeConfiguration";
export * from "./src/doc/handler/ICommandHandler";
export * from "./src/doc/handler/ICommandHandlerResponseChecker";
export * from "./src/doc/handler/ICommandHandlerResponseValidator";
export * from "./src/doc/handler/IHandlerParameters";
export * from "./src/doc/handler/IChainedHandlerEntry";
export * from "./src/doc/handler/IChainedHandlerArgumentMapping";

export * from "./src/doc/option/ICommandOptionAllowableValues";
export * from "./src/doc/option/ICommandOptionDefinition";
export * from "./src/doc/option/ICommandOptionValueImplications";
export * from "./src/doc/option/ICommandPositionalDefinition";

export * from "./src/doc/response/response/ICommandResponse";
export * from "./src/doc/response/parms/ICommandResponseParms";
export * from "./src/doc/response/response/ICommandValidatorError";
export * from "./src/doc/response/response/ICommandValidatorResponse";

export * from "./src/doc/ICommandDefinition";
export * from "./src/doc/ICommandDefinitionPassOn";
export * from "./src/doc/ICommandExampleDefinition";
export * from "./src/doc/IPartialCommandDefinition";

export * from "./src/doc/args/ICommandArguments";

export * from "./src/handlers/FailedCommandHandler";

export * from "./src/help/abstract/AbstractHelpGenerator";
export * from "./src/help/abstract/AbstractHelpGeneratorFactory";

export * from "./src/help/doc/IHelpGeneratorFactory";
export * from "./src/help/doc/IHelpGeneratorFactoryParms";
export * from "./src/help/doc/IHelpGeneratorParms";
export * from "./src/help/DefaultHelpGenerator";
export * from "./src/help/HelpConstants";
export * from "./src/help/HelpGeneratorFactory";
export * from "./src/help/doc/IHelpGenerator";
export * from "./src/help/WebHelpGenerator";
export * from "./src/help/WebHelpManager";

export * from "./src/doc/profiles/definition/ICommandProfile";
export * from "./src/doc/profiles/definition/ICommandProfileTypeConfiguration";

export * from "./src/profiles/CliProfileManager";

export * from "./src/syntax/SyntaxValidator";

export * from "./src/utils/CommandUtils";
export * from "./src/utils/SharedOptions";

export * from "./src/yargs/doc/IYargsParms";
export * from "./src/yargs/doc/IYargsResponse";

export * from "./src/yargs/AbstractCommandYargs";
export * from "./src/yargs/CommandYargs";
export * from "./src/yargs/GroupCommandYargs";
export * from "./src/yargs/YargsConfigurer";
export * from "./src/yargs/YargsDefiner";

export * from "./src/CommandPreparer";
export * from "./src/CommandProcessor";
export * from "./src/response/CommandResponse";

export * from "./src/profiles/CommandProfiles";

export * from "./src/response/CommandResponse";
export * from "./src/response/HandlerResponse";

export * from "./src/doc/response/api/handler/IHandlerResponseApi";
export * from "./src/doc/response/api/handler/IHandlerProgressApi";
export * from "./src/doc/response/api/handler/IHandlerResponseConsoleApi";
export * from "./src/doc/response/api/handler/IHandlerResponseDataApi";
export * from "./src/doc/response/api/handler/IHandlerFormatOutputApi";

export * from "./src/doc/response/response/ICommandOutputFormat";
