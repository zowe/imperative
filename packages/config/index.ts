/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/Config";
export * from "./src/ConfigAutoStore";
export * from "./src/ConfigConstants";
export * from "./src/ConfigSchema";
export * from "./src/ConfigBuilder";
export * as ConfigUtils from "./src/ConfigUtils";
export * from "./src/ProfileCredentials";
export * from "./src/ProfileInfo";
export * from "./src/ProfInfoErr";
export * from "./src/doc/IConfig";
export * from "./src/doc/IConfigAutoStoreOpts";
export * from "./src/doc/IConfigBuilderOpts";
export * from "./src/doc/IConfigConvertResult";
export * from "./src/doc/IConfigLayer";
export * from "./src/doc/IConfigOpts";
export * from "./src/doc/IConfigProfile";
export * from "./src/doc/IConfigSchema";
export * from "./src/doc/IConfigSecure";
export * from "./src/doc/IConfigVault";
export * from "./src/doc/IProfArgAttrs";
export * from "./src/doc/IProfAttrs";
export * from "./src/doc/IProfInfoUpdatePropOpts";
export * from "./src/doc/IProfInfoRemoveKnownPropOpts";
export * from "./src/doc/IProfInfoErrParms";
export * from "./src/doc/IProfLoc";
export * from "./src/doc/IProfMergeArgOpts";
export * from "./src/doc/IProfMergedArg";
export * from "./src/doc/IProfInfoProps";
export * from "./src/doc/IProfOpts";
