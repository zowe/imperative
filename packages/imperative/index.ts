/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/env/EnvironmentalVariableSettings";
export * from "./src/doc/IApimlSvcAttrs";
export * from "./src/doc/IDaemonContext";
export * from "./src/doc/IImperativeEnvironmentalVariableSetting";
export * from "./src/doc/IImperativeEnvironmentalVariableSettings";
export * from "./src/doc/IImperativeConfig";
export * from "./src/doc/IImperativeLoggingConfig";
export * from "./src/ConfigurationLoader";
export * from "./src/ConfigurationValidator";
export * from "./src/DefinitionTreeResolver";
export * from "./src/Imperative";
export * from "./src/LoggingConfigurer";
export * from "./src/plugins/AbstractPluginLifeCycle";
export * from "./src/plugins/PluginManagementFacility";
export * from "./src/auth/doc/IAuthHandlerApi";
export * from "./src/auth/handlers/AbstractAuthHandler";
export * from "./src/auth/handlers/BaseAuthHandler";
export * from "./src/config/cmd/auto-init/handlers/BaseAutoInitHandler";
