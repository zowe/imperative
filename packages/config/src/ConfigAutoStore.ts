/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { ICommandArguments, IHandlerParameters } from "../../cmd";
import { ICommandHandlerRequire } from "../../cmd/src/doc/handler/ICommandHandlerRequire";
import { ICommandProfileAuthConfig } from "../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";
import { IConfigLoadedProfile } from "./doc/IConfigLoadedProfile";
import * as ConfigUtils from "./ConfigUtils";
import { AbstractAuthHandler } from "../../imperative/src/auth/handlers/AbstractAuthHandler";
import { ImperativeConfig } from "../../utilities";
import { ISession } from "../../rest/src/session/doc/ISession";
import { Session } from "../../rest/src/session/Session";
import { AUTH_TYPE_TOKEN } from "../../rest/src/session/SessConstants";
import { Logger } from "../../logger";

/**
 * Class to manage automatic storage of properties in team config.
 */
export class ConfigAutoStore {
    /**
     * Finds the token auth handler class for a team config profile.
     * @param profilePath JSON path of profile
     * @param cmdArguments CLI arguments which may specify a profile
     * @returns Auth handler class or undefined if none was found
     */
    public static findAuthHandlerForProfile(profilePath: string, cmdArguments: ICommandArguments): AbstractAuthHandler | undefined {
        const config = ImperativeConfig.instance.config;
        const profileType = lodash.get(config.properties, `${profilePath}.type`);
        const profile = config.api.profiles.get(profilePath.replace(/profiles\./g, ""));

        if (profile == null || profileType == null) {  // Profile must exist and have type defined
            return;
        } else if (profileType === "base") {
            if (profile.tokenType == null) {  // Base profile must have tokenType defined
                return;
            }
        } else {
            if (profile.basePath == null) {  // Service profiles must have basePath defined
                return;
            } else if (profile.tokenType == null) {  // If tokenType undefined in service profile, fall back to base profile
                const baseProfileName = ConfigUtils.getActiveProfileName("base", cmdArguments);
                return this.findAuthHandlerForProfile(config.api.profiles.expandPath(baseProfileName), cmdArguments);
            }
        }

        const authConfigs: ICommandProfileAuthConfig[] = [];
        ImperativeConfig.instance.loadedConfig.profiles.forEach((profCfg) => {
            if ((profCfg.type === profileType || profCfg.type === "base") && profCfg.authConfig != null) {
                authConfigs.push(...profCfg.authConfig);
            }
        });

        for (const authConfig of authConfigs) {
            const authHandler: ICommandHandlerRequire = require(authConfig.handler);
            const authHandlerClass = new authHandler.default();

            if (authHandlerClass instanceof AbstractAuthHandler) {
                const promptParams = authHandlerClass.getPromptParams()[0];

                if (profile.tokenType === promptParams.defaultTokenType) {
                    return authHandlerClass;  // Auth service must have matching token type
                }
            }
        }
    }

    /**
     * Finds the highest priority layer where a profile is stored.
     * @param loadedProfile
     * @returns User and global properties
     */
    public static getPriorityLayer(loadedProfile: IConfigLoadedProfile): { user: boolean, global: boolean } {
        return {
            user: Object.values(loadedProfile.properties).every(v => v.user),
            global: Object.values(loadedProfile.properties).some(v => v.global)
        };
    }

    /**
     * Stores session config properties into a team config profile.
     * @param params CLI handler parameters object
     * @param sessCfg Session config containing properties to store
     * @param propsToStore Names of properties that should be stored
     */
    public static async storeSessCfgProps(params: IHandlerParameters, sessCfg: { [key: string]: any }, propsToStore: string[]): Promise<void> {
        const config = ImperativeConfig.instance.config;
        // TODO Which autoStore value should take priority if it conflicts between layers
        if (propsToStore.length == 0 || !config?.exists || !config.properties.autoStore) {
            return;
        }

        let profileProps = propsToStore.map(propName => propName === "hostname" ? "host" : propName);
        const profileData = this.findActiveProfile(params, profileProps);
        if (profileData == null) {
            return;
        }
        const [profileType, profileName] = profileData;
        const profilePath = config.api.profiles.expandPath(profileName);

        // Replace user and password with tokenValue if tokenType is defined in config
        if (profileProps.includes("user") && profileProps.includes("password") &&
            await this.fetchTokenForSessCfg(params, sessCfg, profilePath)) {
            profileProps = profileProps.filter(propName => propName !== "user" && propName !== "password");
            profileProps.push("tokenValue");
        }

        const beforeLayer = config.api.layers.get();
        const loadedProfile = config.api.profiles.load(profileName);
        const { user, global } = this.getPriorityLayer(loadedProfile);
        config.api.layers.activate(user, global);

        const baseProfileName = ConfigUtils.getActiveProfileName("base", params.arguments);
        const baseProfileObj = lodash.get(config.properties, config.api.profiles.expandPath(baseProfileName));
        const baseProfileSchema = ImperativeConfig.instance.loadedConfig.baseProfile.schema;
        const profileSchema = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profileType).schema;

        for (const propName of profileProps) {
            let propProfilePath = profilePath;
            // Determine if property should be stored in base profile instead
            if (loadedProfile.properties[propName] == null && !loadedProfile.secure?.includes(propName) &&
                (baseProfileObj.properties[propName] != null || baseProfileObj.secure?.includes(propName) ||
                (propName === "tokenValue" && baseProfileObj.properties.tokenType != null))) {
                propProfilePath = config.api.profiles.expandPath(baseProfileName);
            }
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            const propDefinition = profileSchema.properties[propName] || baseProfileSchema.properties[propName];
            config.set(`${propProfilePath}.properties.${propName}`, sessCfg[sessCfgPropName], {
                secure: propDefinition.secure
            });
        }

        await config.save(false);
        params.response.console.log(`Stored properties in ${config.layerActive().path}: ${profileProps.join(", ")}`);
        // Restore original active layer
        config.api.layers.activate(beforeLayer.user, beforeLayer.global);
    }

    /**
     * Finds the profile where auto-store properties should be saved.
     * @param params CLI handler parameters object
     * @param profileProps List of properties required in the profile schema
     * @returns Tuple containing profile type and name, or undefined if no
     *          profile was found
     */
    private static findActiveProfile(params: IHandlerParameters, profileProps: string[]): [string, string] | undefined {
        const profileTypes = [
            ...(params.definition.profile?.required || []),
            ...(params.definition.profile?.optional || [])
        ];

        for (const profType of profileTypes) {
            const profileMatch = ImperativeConfig.instance.loadedConfig.profiles.find(p => p.type === profType);
            if (profileMatch != null && profileProps.every(propName => propName in profileMatch.schema.properties)) {
                return [profType, ConfigUtils.getActiveProfileName(profType, params.arguments)];
            }
        }
    }

    /**
     * Retrieves token value that will be auto-stored into session config.
     * @param params CLI handler parameters object
     * @param sessCfg Session config with credentials for basic or cert auth
     * @param profilePath JSON path of profile containing tokenType
     * @returns True if auth handler was found and token was fetched
     */
    private static async fetchTokenForSessCfg(params: IHandlerParameters, sessCfg: { [key: string]: any }, profilePath: string): Promise<boolean> {
        const authHandlerClass = this.findAuthHandlerForProfile(profilePath, params.arguments);

        if (authHandlerClass == null) {
            return false;
        }

        const [promptParams, loginHandler] = authHandlerClass.getPromptParams();
        sessCfg.type = AUTH_TYPE_TOKEN;
        sessCfg.tokenType = promptParams.defaultTokenType;
        const baseSessCfg: ISession = { type: sessCfg.type };

        for (const propName of Object.keys(ImperativeConfig.instance.loadedConfig.baseProfile.schema.properties)) {
            const sessCfgPropName = propName === "host" ? "hostname" : propName;
            if (sessCfg[sessCfgPropName] != null) {
                (baseSessCfg as any)[sessCfgPropName] = sessCfg[sessCfgPropName];
            }
        }

        Logger.getAppLogger().info(`Fetching ${sessCfg.tokenType} for ${profilePath}`);
        sessCfg.tokenValue = await loginHandler(new Session(baseSessCfg));
        return true;
    }
}
