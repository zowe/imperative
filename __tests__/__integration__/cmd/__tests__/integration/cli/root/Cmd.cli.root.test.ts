/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITestEnvironment } from "../../../../../../__src__/environment/doc/response/ITestEnvironment";
import { SetupTestEnvironment } from "../../../../../../__src__/environment/SetupTestEnvironment";
import { runCliScript } from "../../../../../../src/TestUtil";
import { ICommandResponse } from "../../../../../../../packages/cmd";
import { Imperative } from "../../../../../../../packages/imperative";

// Test Environment populated in the beforeAll();
let TEST_ENVIRONMENT: ITestEnvironment;

describe("cmd-cli", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await SetupTestEnvironment.createTestEnv({
            cliHomeEnvVar: "CMD_CLI_CLI_HOME",
            testName: "cmd_root"
        });
    });

    it("should display the help", async () => {
        const response = runCliScript(__dirname + "/__scripts__/root_help.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");
        expect(response.stdout.toString()).toMatchSnapshot();
    });

    it("should show a list of available commands", async () => {
        const response = runCliScript(__dirname + "/__scripts__/available_commands.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");
        expect(response.stdout.toString()).toMatchSnapshot();
    });

    it("should flag an invalid group and give a close suggestion", async () => {
        const response = runCliScript(__dirname + "/__scripts__/invalid_command.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);
        expect(response.stdout.toString()).toMatchSnapshot();
        expect(response.stderr.toString()).toMatchSnapshot();
    });

    it("should flag an invalid command and list valid commands", async () => {
        const response = runCliScript(__dirname + "/__scripts__/invalid_command2.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);
        expect(response.stderr.toString()).toContain("Available commands are \"banana-profile, strawberry-profile, kiwi-profile, insecure-profile, base-profile\"");
    });

    it("should display the version", async () => {
        const response = runCliScript(__dirname + "/__scripts__/root_version.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(0);
        expect(response.stderr.toString()).toBe("");
        expect(response.stdout.toString()).toMatchSnapshot();
    });
});
