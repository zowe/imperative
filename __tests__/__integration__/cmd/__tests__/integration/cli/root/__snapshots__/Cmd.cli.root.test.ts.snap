// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cmd-cli should display the help 1`] = `
"================ISSUING HELP NO OPTIONS===============

 DESCRIPTION
 -----------

   A test CLI for the 'cmd' imperative package

 USAGE
 -----

   cmd-cli <group>

   Where <group> is one of the following:

 GROUPS
 ------

   auto-format Invoke handlers to test auto-format           
   chained     chained handler test commands                 
   gen-help    Commands to test help generator               
   invalid     Invalid definitions                           
   invoke      Invoke handlers to test promise reject/fulfill
   nested      Test a complex structure                      
   profile     Validate profile mapping                      
   profiles    Create and manage configuration profiles      
   read        Read some profiles                            
   respond     Invoke handlers that will produce messages    
   validate    Validate syntax checking                      

 OPTIONS
 -------

   --version  | -V (boolean)

      Display the current version of Cmd Package CLI

   --available-commands  | --ac (boolean)

      Displays a list of available commands

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce JSON formatted data from a command

   --help  | -h (boolean)

      Display help text

   --help-web  | --hw (boolean)

      Display HTML help in browser

================ISSUING HELP WITH OPTION==============

 DESCRIPTION
 -----------

   A test CLI for the 'cmd' imperative package

 USAGE
 -----

   cmd-cli <group>

   Where <group> is one of the following:

 GROUPS
 ------

   auto-format Invoke handlers to test auto-format           
   chained     chained handler test commands                 
   gen-help    Commands to test help generator               
   invalid     Invalid definitions                           
   invoke      Invoke handlers to test promise reject/fulfill
   nested      Test a complex structure                      
   profile     Validate profile mapping                      
   profiles    Create and manage configuration profiles      
   read        Read some profiles                            
   respond     Invoke handlers that will produce messages    
   validate    Validate syntax checking                      

 OPTIONS
 -------

   --version  | -V (boolean)

      Display the current version of Cmd Package CLI

   --available-commands  | --ac (boolean)

      Displays a list of available commands

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce JSON formatted data from a command

   --help  | -h (boolean)

      Display help text

   --help-web  | --hw (boolean)

      Display HTML help in browser

================ISSUING HELP WITH RFJ=================
{
  \\"success\\": true,
  \\"exitCode\\": 0,
  \\"message\\": \\"\\",
  \\"stdout\\": \\"\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   A test CLI for the 'cmd' imperative package\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   cmd-cli <group>\\\\n\\\\n   Where <group> is one of the following:\\\\n\\\\n GROUPS\\\\n ------\\\\n\\\\n   auto-format Invoke handlers to test auto-format           \\\\n   chained     chained handler test commands                 \\\\n   gen-help    Commands to test help generator               \\\\n   invalid     Invalid definitions                           \\\\n   invoke      Invoke handlers to test promise reject/fulfill\\\\n   nested      Test a complex structure                      \\\\n   profile     Validate profile mapping                      \\\\n   profiles    Create and manage configuration profiles      \\\\n   read        Read some profiles                            \\\\n   respond     Invoke handlers that will produce messages    \\\\n   validate    Validate syntax checking                      \\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --version  | -V (boolean)\\\\n\\\\n      Display the current version of Cmd Package CLI\\\\n\\\\n   --available-commands  | --ac (boolean)\\\\n\\\\n      Displays a list of available commands\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n   --help-web  | --hw (boolean)\\\\n\\\\n      Display HTML help in browser\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": {}
}"
`;

exports[`cmd-cli should display the version 1`] = `
"================ISSUING VERSION===============
0.0.0
"
`;

exports[`cmd-cli should flag an invalid group and give a close suggestion 1`] = `
"================ISSUING CMD WITH INVALID GROUP===============
"
`;

exports[`cmd-cli should flag an invalid group and give a close suggestion 2`] = `
"Command Error:
Command failed due to improper syntax
Command entered: \\"cmd-cli NOT_REAL\\"
Unknown group: NOT_REAL
Did you mean:  chained?
Use \\"cmd-cli --help\\" to view groups, commands, and options.
"
`;

exports[`cmd-cli should show a list of available commands 1`] = `
"================ISSUING HELP NO OPTIONS===============
cmd-cli auto-format cmd-output

	Allows the user to specify the output format.

cmd-cli auto-format object

	Returns an object.

cmd-cli auto-format table

	Format a table

cmd-cli chained fail-response-mapping

	try to map a non-existent response field, not optional - fail

cmd-cli chained handler-chain-failure

	a handler mid-chain throws an error

cmd-cli chained many-handlers

	many handlers with argument mappings

cmd-cli chained mapping-notexist

	try to map a non-existent response field, optional - succeed

cmd-cli chained optional-response-mapping

	try to map a non-existent response field, optional - succeed

cmd-cli chained print-animals

	print three animal names

cmd-cli chained print-cheetah

	three handlers, but the first two are silent. only prints 'cheetah'

cmd-cli gen-help allowable-options

	Displays allowable options in the help!

cmd-cli gen-help example-test

	Displays example inputs in the help!

cmd-cli invalid invalid-handler

	Test async handler command tests tests a fulfill/reject from an asynchronous
	handler.

cmd-cli invalid no-handler

	This will never get invoked. No handler specified on this definition.

cmd-cli invalid profile-spec

	Command handler attempts to load a profile that wasn't specified on the command
	definition

cmd-cli invoke exit-143

	Test handler that exits with status code 143

cmd-cli invoke test-async-handler

	Test async handler command tests tests a fulfill/reject from an asynchronous
	handler.

cmd-cli invoke test-handler

	Test handler that returns a promise. Tests rejecting and fulfilling the promise
	via the methods.

cmd-cli invoke unexpected-exception-handler

	The unexpected exception handler attempts to perform an operation that will
	result in an exception.

cmd-cli nested first-group first-group-command-one

	A command in the first group

cmd-cli nested first-group first-group-command-two

	A command in the first group

cmd-cli nested first-group second-group second-group-command-one

	A command in the second group

cmd-cli nested first-group second-group second-group-command-two

	A command in the second group

cmd-cli nested first-group third-group third-group-command-one

	A command in the third group

cmd-cli nested nested-group-command

	A command in the nested group

cmd-cli profile mapping

	Tests Imperative's profile to CLI mapping capabilities.

cmd-cli profile mapping-name-type

	Tests that --type and --name are not filled in from profiles.

cmd-cli profile mapping-positional

	Tests Imperative's profile to CLI mapping capabilities.

cmd-cli profiles create banana-profile

	Banana Profile

cmd-cli profiles create insecure-profile

	Test Secured Fields

cmd-cli profiles create strawberry-profile

	Strawberry Profile

cmd-cli profiles delete banana-profile

	Delete a banana profile. You must specify a profile name to be deleted. To find
	a list of available profiles for deletion, issue the profiles list command. By
	default, you will be prompted to confirm the profile removal.

cmd-cli profiles delete insecure-profile

	Delete a insecure profile. You must specify a profile name to be deleted. To
	find a list of available profiles for deletion, issue the profiles list command.
	By default, you will be prompted to confirm the profile removal.

cmd-cli profiles delete strawberry-profile

	Delete a strawberry profile. You must specify a profile name to be deleted. To
	find a list of available profiles for deletion, issue the profiles list command.
	By default, you will be prompted to confirm the profile removal.

cmd-cli profiles list banana-profiles

	Banana Profile

cmd-cli profiles list insecure-profiles

	Test Secured Fields

cmd-cli profiles list strawberry-profiles

	Strawberry Profile

cmd-cli profiles set-default banana-profile

	The banana set default-profiles command allows you to set the default profiles
	for this command group. When a banana command is issued and no profile override
	options are specified, the default profiles for the command group are
	automatically loaded for the command based on the commands profile requirements.

cmd-cli profiles set-default insecure-profile

	The insecure set default-profiles command allows you to set the default profiles
	for this command group. When a insecure command is issued and no profile
	override options are specified, the default profiles for the command group are
	automatically loaded for the command based on the commands profile requirements.

cmd-cli profiles set-default strawberry-profile

	The strawberry set default-profiles command allows you to set the default
	profiles for this command group. When a strawberry command is issued and no
	profile override options are specified, the default profiles for the command
	group are automatically loaded for the command based on the commands profile
	requirements.

cmd-cli profiles update banana-profile

	Banana Profile

cmd-cli profiles update insecure-profile

	Test Secured Fields

cmd-cli profiles update strawberry-profile

	Strawberry Profile

cmd-cli read profile

	Read some profiles

cmd-cli respond with-data-array

	Formulates a string array object to pass back when response format JSON is
	specified.

cmd-cli respond with-data-object

	Formulates a JSON object to pass back when response format JSON is specified.

cmd-cli respond with-error-headers

	Responds with a few error headers (colorized in red).

cmd-cli respond with-error-messages

	Reponds with a few error messages.

cmd-cli respond with-log-messages

	Reponds with a few log messages.

cmd-cli respond with-mixed-responses

	Responds with error, headers, messages, data, etc. (all handler response APIs,
	except progress)

cmd-cli respond with-syntax-errors

	Responds with syntax errors

cmd-cli respond with-wrapped-text

	Responds with wrapped text and non wrapped text

cmd-cli validate string-or-empty

	Test string or empty option.

cmd-cli validate syntax

	Tests Imperatives syntax validation capabilities.

"
`;
